### ISSUE TITLE
/compile:3 does not always compile in memory

### ISSUE THREAD

**Issue by seebees** [dafny-lang/dafny#2861]
```
  /compile:<n>
      0 - Do not compile Dafny program.
      1 (default) - Upon successful verification of the Dafny program,
          compile it to the designated target language. (/noVerify
          automatically counts as a failed verification.)
      2 - Always attempt to compile Dafny program to the target language,
          regardless of verification outcome.
      3 - If there is a Main method and there are no verification errors
          and /noVerify is not used, compiles program in memory (i.e.,
          does not write an output file) and runs it.
      4 - Like (3), but attempts to compile and run regardless of
          verification outcome.
```

However, `go`, `java` and `py` compileTarget all generate extensive files on disk.
We should either not write to disk or update the option.

I also think it would be better to be consistent here.
So if some runtimes require writing to disk,
do it for every runtime.

**Comment by robin-aws** [comment]
I agree with updating the option. It was correct when C# was the only option but hasn't been updated. There are probably a few other options that need a similar update.

Re: writing to disk, I'd go in the other direction to achieve consistency: if writing to disk is required, it should be done in a temporary directory rather than right in the current directory.

This can be cleaner with the new CLI commands, since having to write to disk somewhere is implementation detail for `build` and `run`, but part of the actual API for `translate`.

**Comment by seebees** [comment]
Then you should also know that the `-out` option controls where the temp files go.

I suppose that the principle of "customers never see it" is an ok alternative as well.
The difference is unlikely to matter for build systems.

**Comment by keyboardDrummer** [comment]
I suggest that even for C# it does not compile in-memory, but instead invokes a separate dotnet process based on a .csproj file. This'll make it easier to debug issues.
